### PR TITLE
bash: avoid using unspliced dependency

### LINKED ISSUE #243164: `pkgsCross.gnu64.bash` fails to build on `x86_64-linux`
### Describe the bug

At current `master`, this fails:

```
nix build pkgsCross.gnu64.bash --system x86_64-linux
```

### Screenshots
```
$ git rev-parse HEAD
ba448b056b5dcd97eb875596a0962912771a2bdf

$ nix build -f . -L pkgsCross.gnu64.bash
bash> configure flags: --disable-static --prefix=/nix/store/cygrphwqmv7lmcgdk53m76sw4zp8c5f0-bash-5.2-p15-x86_64-unknown-linux-gnu --bindir=/nix/store/cygrphwqmv7lmcgdk53m76sw4zp8c5f0-bash-5.2-p15-x86_64-unknown-linux-gnu/bin --sbindir=/nix/store/cygrphwqmv7lmcgdk53m76sw4zp8c5f0-bash-5.2-p15-x86_64-unknown-linux-gnu/sbin --includedir=/nix/store/fq5ifn83r9y7qzfrabj2pcsqj8i5mri2-bash-5.2-p15-x86_64-unknown-linux-gnu-dev/include --oldincludedir=/nix/store/fq5ifn83r9y7qzfrabj2pcsqj8i5mri2-bash-5.2-p15-x86_64-unknown-linux-gnu-dev/include --mandir=/nix/store/bbvqpyzvxf92spv70ds2dah3fjkaxwi5-bash-5.2-p15-x86_64-unknown-linux-gnu-man/share/man --infodir=/nix/store/j957krj8qlfzawp30i45w99lr0jrxzm9-bash-5.2-p15-x86_64-unknown-linux-gnu-info/share/info --docdir=/nix/store/rwmvr3pc2r7hfhzny28yhdr0vaq2zamx-bash-5.2-p15-x86_64-unknown-linux-gnu-doc/share/doc/bash --libdir=/nix/store/cygrphwqmv7lmcgdk53m76sw4zp8c5f0-bash-5.2-p15-x86_64-unknown-linux-gnu/lib --libexecdir=/nix/store/cygrphwqmv7lmcgdk53m76sw4zp8c5f0-bash-5.2-p15-x86_64-unknown-linux-gnu/libexec --localedir=/nix/store/cygrphwqmv7lmcgdk53m76sw4zp8c5f0-bash-5.2-p15-x86_64-unknown-linux-gnu/share/locale --disable-readline bash_cv_job_control_missing=nomissing bash_cv_sys_named_pipes=nomissing bash_cv_getcwd_malloc=yes --build=x86_64-unknown-linux-gnu --host=x86_64-unknown-linux-gnu
bash> configure: WARNING: unrecognized options: --disable-static
bash> checking build system type... x86_64-unknown-linux-gnu
bash> checking host system type... x86_64-unknown-linux-gnu
bash> Beginning configuration for bash-5.2-release for x86_64-unknown-linux-gnu
bash> checking for x86_64-unknown-linux-gnu-gcc... x86_64-unknown-linux-gnu-gcc
bash> checking whether the C compiler works... no
bash> configure: error: in `/build/bash-5.2':
bash> configure: error: C compiler cannot create executables
bash> See `config.log' for more details
```

From `config.log`:

```
configure:4363: $? = 1
configure:4383: checking whether the C compiler works
configure:4405: x86_64-unknown-linux-gnu-gcc    conftest.c  >&5
/nix/store/zsvpmlddl2i5mpzm031a99xfpn410m5b-binutils-2.40/bin/ld: cannot find crt1.o: No such file or directory
/nix/store/zsvpmlddl2i5mpzm031a99xfpn410m5b-binutils-2.40/bin/ld: cannot find crti.o: No such file or directory
/nix/store/zsvpmlddl2i5mpzm031a99xfpn410m5b-binutils-2.40/bin/ld: cannot find -lgcc_s: No such file or directory
collect2: error: ld returned 1 exit status
```

### PR DESCRIPTION
**Pull request by ghost**
## Description of changes

`stdenv` is spliced, but `stdenv.cc` is not, so we can't use the
latter in dependencies.

Closes

- https://github.com/NixOS/nixpkgs/issues/243164

Includes

- https://github.com/NixOS/nixpkgs/pull/264976

Fixes #243164

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by alyssais** [comment]
> `stdenv` is spliced, but `stdenv.cc` is not, so we can't use the
> latter in dependencies.

I'm a bit confused what you mean by this — are you saying `buildPackages.stdenv.cc` wasn't getting us what we expected?  Should it just have been `pkgsBuildBuild.stdenv.cc`, since it was in `depsBuildBuild`?

**Comment by amjoseph-nixpkgs** [comment]
> I'm a bit confused what you mean by this — are you saying `buildPackages.stdenv.cc` wasn't getting us what we expected?

Correct.

Try changing `buildPackages.stdenv.cc` to `assert buildPackages.stdenv.cc?__spliced; buildPackages.stdenv.cc` and you'll get an assertion.

> Should it just have been `pkgsBuildBuild.stdenv.cc`, since it was in `depsBuildBuild`?

No; you should not use `pkgsFooBar` anywhere in `depsFooBar` (or `buildInputs` or `nativeBuildInputs`) because `pkgsFooBar` are unspliced packages.

See:

- https://github.com/NixOS/nixpkgs/pull/263082

which adds a `lib.warn` when people do this.  I hope to merge that right after 23.11 branch-off.

File: pkgs/shells/bash/5.nix [hunk 1]
```diff
@@ -91,7 +91,7 @@ stdenv.mkDerivation rec {
 
   strictDeps = true;
   # Note: Bison is needed because the patches above modify parse.y.
-  depsBuildBuild = [ buildPackages.stdenv.cc ];
+  depsBuildBuild = [ stdenv ];
```
Comment (Artturin) [inline pkgs/shells/bash/5.nix:94]: I don't think this will work? how would `stdenv` become `stdenv.cc`?

Comment (amjoseph-nixpkgs) [reply pkgs/shells/bash/5.nix:94]: You're right; this works "by accident".

**Comment by amjoseph-nixpkgs** [comment]
This only works "by accident".

**Pull request closed without merging**